File: cambiandosedeaula.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <tuple>
#include <string>
#include <cstdlib>

using namespace std;

struct Key {
    int i, j, suma;

    bool operator==(const Key &other) const {
        return i == other.i && j == other.j && suma == other.suma;
    }
};

struct KeyHash {
    size_t operator()(const Key &k) const {
        return hash<int>()(k.i) ^ hash<int>()(k.j << 1) ^ hash<int>()(k.suma << 2);
    }
};

bool caminoMinimo(int i, int j, int suma,
                  const vector<vector<int>> &matriz,
                  int filas, int cols,
                  unordered_map<Key, bool, KeyHash> &memo) {
    if (i >= filas || j >= cols)
        return false;

    suma += matriz[i][j];

    if (abs(suma) > filas + cols)
        return false;

    if (i == filas - 1 && j == cols - 1)
        return suma == 0;

    Key k{i, j, suma};
    if (memo.find(k) != memo.end())
        return memo[k];

    bool res = caminoMinimo(i + 1, j, suma, matriz, filas, cols, memo) ||
               caminoMinimo(i, j + 1, suma, matriz, filas, cols, memo);

    memo[k] = res;
    return res;
}

int main() {
    int cantidadDeCasos;
    cin >> cantidadDeCasos;

    while (cantidadDeCasos--) {
        int filas, cols;
        cin >> filas >> cols;

        vector<vector<int>> matriz(filas, vector<int>(cols));
        for (int i = 0; i < filas; ++i) {
            for (int j = 0; j < cols; ++j) {
                cin >> matriz[i][j];
            }
        }

        if ((filas % 2 == 0 && cols % 2 == 0) || (filas % 2 != 0 && cols % 2 != 0)) {
            cout << "NO" << endl;
            continue;
        }

        unordered_map<Key, bool, KeyHash> memo;
        if (caminoMinimo(0, 0, 0, matriz, filas, cols, memo)) {
            cout << "YES" << endl;
        } else {
            cout << "NO" << endl;
        }
    }

    return 0;
}

File: cambiandosedeaula.py
def cambiandoseDeAula ():
    cantidad_de_casos = int(input())
    resultados = []

    for _ in range(cantidad_de_casos):
        filas, cols = map(int, input().split())

        matriz = []
        for _ in range(filas):
            fila = list(map(int, input().split()))
            matriz.append(fila)

        if filas % 2 == 0 and cols % 2 == 0: 
            resultados.append("NO")
            continue
        elif filas % 2 != 0 and cols % 2 != 0: 
            resultados.append("NO")
            continue

        memo = {}
        def caminimoMinimo (i, j, suma):
            if i >= filas or i < 0 or j < 0 or j >= cols: # para que no se pasen los indices
                return False

            suma += matriz[i][j] # Voy sumando los weights

            if abs(suma) > filas + cols: return False 

            if i == filas - 1 and j == cols - 1: # Si llegamos al extremo inferior derecho me fijo si la suma acumulada da 0
                if suma == 0: return True
                else: return False

            clave_memo = (i, j, suma)
            if clave_memo in memo: return memo[clave_memo] # Si tengo la clave, devuelvo su valor guardado

            res = caminimoMinimo(i+1, j, suma) or caminimoMinimo(i, j+1, suma) # O voy a la derecha o a la izquierda

            memo[clave_memo] = res
            return res

        if caminimoMinimo(0,0,0): resultados.append("YES")
        else: resultados.append("NO")
        
    for resultado in resultados: 
        print(resultado)

cambiandoseDeAula()